File: packages/validator-go/validator/validate/rule_length.go
package validate

// Length rules minlength, maxlength and rangelength (docs/spec/validation-rules.md,
// Values): they count the code points of a scalar's canonical text, untrimmed; an
// array or object value fails. Their parameters are checked by checkLengthLimit
// and checkLengthRange before a rule runs.

import (
	"math"
	"strings"
	"unicode/utf8"
)

// maxLengthLimit is the largest length limit, 2^53 - 1.
const maxLengthLimit = 9007199254740991

// lengthLimit returns a length limit: an integer from 0 to maxLengthLimit.
func lengthLimit(param any) (float64, bool) {
	n, ok := numberValue(param)
	if !ok || math.IsNaN(n) || n < 0 || n > maxLengthLimit || n != math.Trunc(n) {
		return 0, false
	}
	return n, true
}

// lengthRange returns the [minimum, maximum] limits of rangelength.
func lengthRange(param any) (float64, float64, bool) {
	limits, ok := param.([]any)
	if !ok || len(limits) != 2 {
		return 0, 0, false
	}
	lo, okLo := lengthLimit(limits[0])
	hi, okHi := lengthLimit(limits[1])
	if !okLo || !okHi || lo > hi {
		return 0, 0, false
	}
	return lo, hi, true
}

// checkLengthLimit returns the parameter error of a minlength or maxlength limit.
func checkLengthLimit(rule string, param any) *parameterError {
	if _, ok := lengthLimit(param); ok {
		return nil
	}
	return ruleParameterError("Invalid " + rule + " parameter: expected an integer from 0 to 9007199254740991")
}

// checkLengthRange returns the parameter error of rangelength limits.
func checkLengthRange(param any) *parameterError {
	if _, _, ok := lengthRange(param); ok {
		return nil
	}
	return ruleParameterError("Invalid rangelength parameter: expected [minimum, maximum] integers with minimum not above maximum")
}

// textLength returns the code-point length of a value's canonical text; false
// when the value has none (an array, an object or a nonfinite number).
func textLength(value any) (float64, bool) {
	text, ok := canonicalText(value)
	if !ok {
		return 0, false
	}
	return float64(utf8.RuneCountInString(text)), true
}

// ruleMinLength fails a value shorter than the limit. Empty values pass.
func ruleMinLength(value any, ruleParam any, ctx ruleContext) (string, bool) {
	if isEmpty(value) {
		return "", false
	}
	limit, _ := lengthLimit(ruleParam)
	if length, ok := textLength(value); !ok || length < limit {
		m := ctx.msg("minlength", "Please enter at least {0} characters.")
		return strings.ReplaceAll(m, "{0}", canonicalNumber(limit)), true
	}
	return "", false
}

// ruleMaxLength fails a value longer than the limit. Empty values pass.
func ruleMaxLength(value any, ruleParam any, ctx ruleContext) (string, bool) {
	if isEmpty(value) {
		return "", false
	}
	limit, _ := lengthLimit(ruleParam)
	if length, ok := textLength(value); !ok || length > limit {
		m := ctx.msg("maxlength", "Please enter no more than {0} characters.")
		return strings.ReplaceAll(m, "{0}", canonicalNumber(limit)), true
	}
	return "", false
}

// ruleRangeLength fails a value whose length is outside [minimum, maximum].
// Empty values pass.
func ruleRangeLength(value any, ruleParam any, ctx ruleContext) (string, bool) {
	if isEmpty(value) {
		return "", false
	}
	lo, hi, _ := lengthRange(ruleParam)
	if length, ok := textLength(value); !ok || length < lo || length > hi {
		m := ctx.msg("rangelength", "Please enter a value between {0} and {1} characters.")
		m = strings.ReplaceAll(m, "{0}", canonicalNumber(lo))
		m = strings.ReplaceAll(m, "{1}", canonicalNumber(hi))
		return m, true
	}
	return "", false
}

File: packages/php-ext/src/rule_length.c
#include "engine_internal.h"

#include <math.h>

/*
 * Length rules (docs/spec/validation-rules.md, "Values"): minlength, maxlength and rangelength
 * count the code points of a scalar's canonical text, untrimmed; an array or object fails.
 */

#define LENGTH_LIMIT_MAXIMUM INT64_C(9007199254740991)

bool ps_length_limit(const ps_value *parameter, int64_t *limit)
{
    if (!parameter) return false;
    if (parameter->kind == PS_INT) {
        if (parameter->data.integer < 0 || parameter->data.integer > LENGTH_LIMIT_MAXIMUM) return false;
        *limit = parameter->data.integer;
        return true;
    }
    if (parameter->kind == PS_FLOAT) {
        double number = parameter->data.number;
        if (!isfinite(number) || floor(number) != number || number < 0 || number > (double)LENGTH_LIMIT_MAXIMUM)
            return false;
        *limit = (int64_t)number;
        return true;
    }
    return false;
}

bool ps_length_range(const ps_value *parameter, int64_t *minimum, int64_t *maximum)
{
    return parameter && parameter->kind == PS_ARRAY && ps_size(parameter) == 2 &&
        ps_length_limit(ps_at(parameter, 0), minimum) &&
        ps_length_limit(ps_at(parameter, 1), maximum) && *minimum <= *maximum;
}

bool ps_length_rule(ps_text rule)
{
    return ps_text_is(rule, "minlength") || ps_text_is(rule, "maxlength") || ps_text_is(rule, "rangelength");
}

ps_parameter_problem ps_length_parameter(ps_text rule, const ps_value *parameter)
{
    int64_t minimum, maximum;
    if (ps_text_is(rule, "rangelength")) {
        if (ps_length_range(parameter, &minimum, &maximum)) return (ps_parameter_problem){0};
        return (ps_parameter_problem){.code = "INVALID_RULE_PARAMETER", .message =
            "Invalid rangelength parameter: expected [minimum, maximum] integers with minimum not above maximum"};
    }
    if (ps_length_limit(parameter, &minimum)) return (ps_parameter_problem){0};
    return (ps_parameter_problem){.code = "INVALID_RULE_PARAMETER", .message =
        ps_text_is(rule, "minlength")
            ? "Invalid minlength parameter: expected an integer from 0 to 9007199254740991"
            : "Invalid maxlength parameter: expected an integer from 0 to 9007199254740991"};
}

int ps_length_passes(ps_text rule, const ps_value *value, const ps_value *parameter)
{
    int64_t minimum = 0, maximum = LENGTH_LIMIT_MAXIMUM;
    if (ps_text_is(rule, "rangelength")) {
        if (!ps_length_range(parameter, &minimum, &maximum)) return -1;
    } else {
        int64_t limit;
        if (!ps_length_limit(parameter, &limit)) return -1;
        if (ps_text_is(rule, "minlength")) minimum = limit;
        else maximum = limit;
    }
    ps_chars text;
    int scalar = ps_canonical_text(value, &text);
    if (scalar <= 0) return scalar;
    size_t count = ps_code_points(ps_view(text));
    free(text.bytes);
    return (uint64_t)count >= (uint64_t)minimum && (uint64_t)count <= (uint64_t)maximum;
}
